### PR TITLE
clientconn: enforce dial timeout in addrConn.wait()

### PR DESCRIPTION
**Pull request by heyitsanthony**
Otherwise a blocking rpc will wait forever for a downed endpoint, even if
there are connected endpoints available from the balancer.

/cc @iamqizhao @xiang90

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -491,6 +495,9 @@ func (cc *ClientConn) getTransport(ctx context.Context, opts BalancerGetOptions)
 		cc.mu.RLock()
 		if cc.conns == nil {
 			cc.mu.RUnlock()
+			if put != nil {
+				put()
```
Comment (xiang90) [inline clientconn.go]: is this relevant to this commit?

Comment (heyitsanthony) [inline clientconn.go]: It's related to the connection path making sense, but not this particular commit. Reverted.

**Comment by xiang90** [comment]
@iamqizhao @menghanl This is a minimal change, can you take a look?

**Comment by menghanl** [comment]
Can you specify a timeout for the blocking RPC, so it will return deadline exceeded error?

Reusing dial timeout seems not to be a good idea. `ac.wait` should not exit on dial timeout.
With this change, the real timeout for each RPC becomes `min(context-timeout, dial-timeout)`.

**Comment by menghanl** [comment]
The issue with `put` seems to be a real one. I will have a PR for that.

**Comment by heyitsanthony** [comment]
@menghanl are you suggesting putting the timeout in the rpc ctx? That's no better than FailFast. I want to submit a blocking RPC such that if the first selected endpoint is down, then grpc will select a new one. The alternative is wrapping every RPC call in retry logic that, as far as I can tell, belongs in the transport layer. Using the dial timeout for selecting a new endpoint makes sense since waiting for the transport is essentially waiting for a dial to complete.

**Comment by heyitsanthony** [comment]
@menghanl @iamqizhao this is my workaround: https://github.com/coreos/etcd/pull/5845/commits/e4946c964c597dcd2c3d7bb7301deed96ae2d696 so that blocking grpc failover will even work using a balancer. It's still incomplete because it needs the wrapper to be plumbed through everything so that the dial timeout is propagated from the client configuration.

Similarly, it's still inferior to handling it in grpc because if rpcTime > dialTimeout, it'll always fail even if a connection is available (this is not the case if the timeout is in `wait()`). Presumably one could inspect the incoming ctx string to get the timeout then take the max, but that is still unsatisfactory for the regular blocking / no timeout case because the dial timeout would fail it too early.

From a broader scope, anything that wants to use the raw grpc client interface will also have to know about the timeout/retry wrapper instead of depending on the grpc connection to recognize the endpoint is down and get another address.

Am I doing something wrong? This workaround is objectively garbage and I'd rather not ruin perfectly good code doing it this way if it's not necessary.

**Comment by menghanl** [comment]
@heyitsanthony If a connection error happens when client is writing or reading, the RPC (unary) will retry on other available connections.
It seems there is a race between balancer and gRPC internal. The connection is up when calling `balancer.Get()`, but down immediately after that before writing anything.
Will think about how to resolve that.

**Comment by xiang90** [comment]
> If a connection error happens when client is writing or reading, the RPC (unary) will retry on other available connections.

In some cases, we do not want the RPC to be retried after finishing writing. Basically, at that point we do not know if the RPC succeeds or now. Retrying write request can have bad side effect.

**Comment by menghanl** [comment]
Sorry I forgot to mention, I was talking about non-failfast RPCs. Failfast RPCs will fail on connection errors.

**Comment by xiang90** [comment]
> Sorry I forgot to mention, I was talking about non-failfast RPCs. Failfast RPCs will fail on connection errors.

So for non-retryable RPC, we need to use failfast option?

**Comment by menghanl** [comment]
Yes, failfast RPCs (unary) will not retry when connection error happens.

**Comment by xiang90** [comment]
@menghanl 

Ideally, we want gRPC to retry when there is a connection error in the dial phase or before the request finishes sending, and not to retry when the request is already sent. Basically we want at most once semantics working with endpoint choosing/dialing.

Is it possible today? Or we cannot use the balancer.

**Comment by heyitsanthony** [comment]
> With this change, the real timeout for each RPC becomes min(context-timeout, dial-timeout).

OK, it looks like all that had to change is `getTransport` callers need to recognize `ErrClientConnTimeout` so that they retry on `!FailFast`. Updated PR to reflect this.

**Comment by tamird** [comment]
+1 that retries should happen in case of failure before transmission of the RPC. please do not violate at-most-once.

**Comment by menghanl** [comment]
@heyitsanthony What I'm thinking now is that maybe non-failfast RPCs should not wait (even for dial timeout) on the down-ed connection. It should restart immediately and get a new connection. I will try and see if that's reasonable. Will keep you informed.

**Comment by iamqizhao** [comment]
@xiang90 Sorry about the delay. Just back from the vacation. I see two questions in this thread:
i) A connection error happens between Balancer.Get returns and addrConn.Wait starts. @menghanl just sent #830 to fix the issue. Note that right dial timeout only applies to the initial dialing and therefore should not be applied here;
ii) You want to retry an RPC when there is a connection error in the dial phase or before the request finishes sending, and not to retry when the request is already sent. This lives in the position between fail-fast and non-failfast we defined. I would like to emphasize that even though grpc write returns successfully there is no guarantee your request msg can reach the server (it still could fail on any part of the path client host <--> network <--> server host). Receiving the response is the only way to justify whether your request is received by the server. Therefore, I feel the retry scenario you described is somewhat problematic (I could be wrong because I am not clear about your use case :). I guess your decision is from the fact that your server is stateful (e.g., a successful request could acquire a file lock already.). According to my experience working with Chubby in Google, you need a special rpc to roll back any potential operations which have been done on the server side when one of your rpcs does not complete successfully.

**Comment by xiang90** [comment]
>  I would like to emphasize that even though grpc write returns successfully there is no guarantee your request msg can reach the server (it still could fail on any part of the path client host <--> network <--> server host).

I understand this. And this is what I would expect.

Here is our use case:

Write requests have side-effect. We cannot retry on any write requests that potentially succeeds (already wrote out, waiting for reply). 

We do want to retry write request if there is a connection error. Say users give 3 endpoints, and 1 is bad. If the write goes to the bad one, we should retry.

In my opinion, each RPC can have three state:

1. waiting for get an endpoint from balancer
2. waiting for dialing the endpoint (we might ignore this if the returned endpoint is connected)
3. writing the requests
4. waiting for the response
5. finishing up

Basically I hope the retry happens before we entry state 3, but not after. fast-fail will return from state 2 without retry.

**Comment by iamqizhao** [comment]
I think even if grpc makes this change for you you still need to deal with the same problem if write returns successfully in step 3 but there is an kernel/network error so that both you and grpc have no idea whether the server has received the request. In my understanding, your proposal just shrink the retry corner a bit but you still need to deal with all the problems without this change.

**Comment by xiang90** [comment]
@iamqizhao 

You are right that someone has to handle failures. But for us, we write a client that using grpc-go. For any error happens after state 2 (successfully dialed), we can directly return them to our users since etcd does not know what happened either just like grpc. For error happens before state 2 (including), we need to handle it. It is unreasonable for us to return an error to our users if it is just a connection failure. 

My general feeling is that for the things that we are certain, we should retry. Or every libraries that use grpc-go will have to implement the similar logic.

**Comment by menghanl** [comment]
I think this is already fixed.
I will double check and get back to this.

**Comment by menghanl** [comment]
I believe the original problem that RPCs block on downed connections has be resolved in #830.

On the discussions about retrying, we have a new implementation on retry this coming quarter.

Closing this PR now.
